stdlib/2/typing.pyi

```diff
@@ -352,64 +351,77 @@ class TextIO(IO[unicode]):
 class ByteString(Sequence[int]): ...
 
 class Match(Generic[AnyStr]):
-    pos = 0
-    endpos = 0
-    lastindex = 0
-    lastgroup = ...  # type: AnyStr
-    string = ...  # type: AnyStr
+    pos: int
+    endpos: int
+    lastindex: Optional[int]
+    string: AnyStr
 
     # The regular expression object whose match() or search() method produced
-    # this match instance.
-    re = ...  # type: 'Pattern[AnyStr]'
+    # this match instance. This should not be Pattern[AnyStr] because the type
+    # of the pattern is independent of the type of the matched string in
+    # Python 2. Strictly speaking Match should be generic over AnyStr twice:
+    # once for the type of the pattern and once for the type of the matched
+    # string.
+    re: Pattern[Any]
+    # Can be None if there are no groups; otherwise matches the type of the pattern.
+    lastgroup: Any
 
-    def expand(self, template: AnyStr) -> AnyStr: ...
+    def expand(self, template: Union[str, Text]) -> AnyStr: ...
```

Comment:
I don't think this is right. Experimentally the return type can be unicode when the template is unicode even when the pattern is not:
```py
>>> re.compile(r'\s+(\w+)[^\w\s]*\s+').match(' foo ').expand(r'(\1)')
'(foo)'
>>> re.compile(r'\s+(\w+)[^\w\s]*\s+').match(u' foo ').expand(r'(\1)')
u'(foo)'
>>> re.compile(r'\s+(\w+)[^\w\s]*\s+').match(' foo ').expand(ur'(\1)')
u'(foo)'
```

Reply:
Sorry for missing this comment. Perhaps we should just use `Any` here then in the return type.